[wasm] Add support for using emscripten from packages

radical opened the pull request:
When using a workload that has the emscripten packages, certain
properties get set. If that isn't set then we fallback to using
`$(EMSDK_PATH)`.

This should allow being able to use emscripten from workloads, but for
other cases where we depend on an installed emscripten, that should work
too.

radical committed:
[wasm] Add support for using emscripten from packages

When using a workload that has the emscripten packages, certain
properties get set. If that isn't set then we fallback to using
`$(EMSDK_PATH)`.

This should allow being able to use emscripten from workloads, but for
other cases where we depend on an installed emscripten, that should work
too.

radical committed:
Fix setting PATH on windows

radical committed:
Using msbuild NormalizeDirectory to construct dir path

lewing commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -72,13 +71,23 @@
+    <EmscriptenSdkToolsPath>$([MSBuild]::EnsureTrailingSlash($(EMSDK_PATH)))</EmscriptenSdkToolsPath>
+    <EmSdkUpstreamBinPath>$([MSBuild]::NormalizeDirectory($(EmscriptenSdkToolsPath), 'upstream', 'bin'))</EmSdkUpstreamBinPath>
+
+    <EmscriptenNodeToolsPath>$([MSBuild]::NormalizeDirectory($(EmscriptenSdkToolsPath), 'node', '12.18.1_64bit'))</EmscriptenNodeToolsPath>
+    <EmscriptenPythonToolsPath>$([MSBuild]::NormalizeDirectory($(EmscriptenSdkToolsPath), 'python', '3.7.4-2_64bit'))</EmscriptenPythonToolsPath>
```
the versions should probably explicit props and a check to make sure it exists

radical replied:
This is closely tied to Emscripten 2.0.12, so the paths won't change separately I think. I'll add the Exists checks.

lewing replied:
That was the point I was making, they version with the emsdk version so let's make sure the build breaks in a way that is obvious

radical replied:
And having the versions there at all caused an issue:
- CI has Emscripten 2.0.14
- The packages that I was testing with was 2.0.12

.. which had differing python/node versions, .. and that caused the build to fail!

lewing reviewed (approved):
I think this looks reasonable.  I wish we had windows AOT tests running

radical committed:
fix setting PATH on windows, when paths have trailing slash

radical committed:
[wasm] Improve checking emscripten from workload, and EMSDK_PATH

- If a workload pack is being used, then ignore `EMSDK_PATH`
- And error in that order too
- Emit better error messages

Example messages:
- Missing python dir when relinking:

`warning : Specified Emscripten sdk at $(EMSDK_PATH)=. is missing some paths: $(EmscriptenPythonToolsPath)=/Users/radical/dev/r3/artifacts/bin/Wasm.Build.Tests/net6.0-Release/browser-wasm/e10djxv0.wdr/python/3.7.4-2_64bit/ . Emscripten SDK is required for building native files.`

- Missing node directory, or the wrong version dir:

`warning : Specified Emscripten sdk at $(EMSDK_PATH)=/Users/radical/dev/r3/src/mono/wasm/emsdk is missing some paths: $(EmscriptenNodeToolsPath)=/Users/radical/dev/r3/src/mono/wasm/emsdk/node/12.18.1_64bit/ . Emscripten SDK is required for building native files.`

- Empty `EMSDK_PATH`

`Error : Could not find Emscripten sdk at $(EMSDK_PATH)=/Users/radical/dev/r3/src/mono/wasm/emsdk/. Emscripten SDK is required for AOT'ing assemblies.`

- Bad `EMSDK_PATH`

`error : Specified Emscripten sdk at $(EMSDK_PATH)=. is missing some paths: $(EmscriptenPythonToolsPath)=/Users/radical/dev/r3/artifacts/bin/Wasm.Build.Tests/net6.0-Release/browser-wasm/l2mcsipf.isn/python/3.7.4-2_64bit/ . Emscripten SDK is required for AOT'ing assemblies.`

radical committed:
[wasm] Don't hardcode versions for python/node bundled in emsdk

.. and listen to @lewing's suggestions!

radical committed:
[wasm] GetDirectories doesn't like non-existant paths

radical committed:
Fix the earlier fix

radical committed:
improve errors when some paths are missing

radical committed:
cleanup

radical committed:
debugging on ci

radical committed:
[wasm] emscripten uses system python on linux, so don't try to find it in emscripten sdk

lewing commented:
/backport to release/6.0-preview3
